test/bitcoin/bitcoinTestTools.mo

```diff
@@ -0,0 +1,140 @@
+import Curves "../../src/ec/Curves";
+import Hash "../../src/Hash";
+import Fp "../../src/ec/Fp";
+import Types "../../src/bitcoin/Types";
+import PublicKey "../../src/ecdsa/Publickey";
+import Common "../../src/Common";
+import Debug "mo:base/Debug";
+import Array "mo:base/Array";
+import Nat8 "mo:base/Nat8";
+import Iter "mo:base/Iter";
+
+
+module {
+  public type Signature = {r : Nat; s : Nat};
+  let curve = Curves.secp256k1;
+
+  public class Account(_sk : Nat) {
+    public let sk : Nat = _sk;
+    public let point = Jacobi.toAffine(Jacobi.mulBase(sk, Curves.secp256k1));
+    // public let address = Address.keyToP2pkh(#test, Affine.toBytes(point, true));
+    public let pkData = Affine.toBytes(point, true);
+    public let pkhash = Hash.hash160(pkData);
+    public let pk = switch(PublicKey.fromPoint(point)) {
+      case (#ok(pk)) {
+        pk;
+      };
+      case (_) {
+        Debug.trap("Invalid");
+      }
+    };
+  };
+
+  func Fr(value : Nat) : Fp.Fp {
+    return Fp.Fp(value, curve.r);
+  }; 
+
+  public func ecdsaSign(sk : Nat, rand : Nat, message : [Nat8]) : Signature {
+    let h = Common.readBE256(Hash.doubleSHA256(message), 0);
+    switch(Jacobi.toAffine(Jacobi.mulBase(rand, Curves.secp256k1))) {
+      case (#point (x, y, curve)) {
+        let r = x.value;
+        if (r == 0) {
+          Debug.trap("r = 0, use different rand.");
+        };
+        let s = Fr(rand).inverse().mul(
+          Fr(h + sk * r)
+        );
+        if (s.value == 0) {
+          Debug.trap("s = 0, use different rand.");
+        };
+
+        let finalS = if (s.value > curve.r/2) {
+          curve.r - s.value
+        } else {
+          s.value
+        };
+
+        return {r = r; s = finalS};
+      };
+      case (#infinity (_)) {
+        Debug.trap("Computed infinity point, use different rand.");
+      };
+    };
+  };
+
+  public func signatureToDer(signature : Signature,
+    sighashType : Types.SighashType) : [Nat8] {
+    // 0x30 [total-length] 0x02 [R-length] [R] 0x02 [S-length] [S] [sighash-type]
+
+    func prepSignatureMember(value : Nat) : ([Nat8], Nat, Nat, Bool) {
+      let data = Array.init<Nat8>(32, 0);
+      Common.writeBE256(data, 0, value);
+      var startOffset = 0;
+      label L for (i in Iter.range(0, data.size() - 1)) {
+        if (data[i] == 0) {
+          startOffset += 1;
+        } else {
+          break L;
+        };
+      };
+      let prependZero : Bool = data[startOffset] >= 0x80;
+      let totalSize = if (prependZero) {
+        data.size() - startOffset + 1
+      } else {
+        data.size() - startOffset
+      };
+      return (Array.freeze(data), startOffset, totalSize, prependZero);
+    };
+
+    let (rData, rStartOffset, rTotalSize,
+      rPrependZero) = prepSignatureMember(signature.r);
+
+    let (sData, sStartOffset, sTotalSize,
+      sPrependZero) = prepSignatureMember(signature.s);
+
+    let totalSize = 1 + 1 + 1 + 1 + rTotalSize + 1 + 1 + sTotalSize + 1;
```

Comment:
Can you elaborate on how you get this total size?

Reply:
Done